samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -0,0 +1,319 @@
+                m -> m,
+                KVSerde.of(keySerde, relMsgSerde),
+                "stream_" + joinId)
+            .map(KV::getValue)
+            .join(table, joinFn);
+
+    context.registerMessageStream(join.getId(), outputStream);
+  }
+
+  private void validateJoinQuery(LogicalJoin join) {
+    JoinRelType joinRelType = join.getJoinType();
+
+    if (joinRelType.compareTo(JoinRelType.INNER) != 0 && joinRelType.compareTo(JoinRelType.LEFT) != 0
+        && joinRelType.compareTo(JoinRelType.RIGHT) != 0) {
+      throw new SamzaException("Query with only INNER and LEFT/RIGHT OUTER join are supported.");
+    }
+
+    boolean isTablePosOnLeft = isTable(join.getLeft());
+    boolean isTablePosOnRight = isTable(join.getRight());
+
+    if (!isTablePosOnLeft && !isTablePosOnRight) {
+      throw new SamzaException("Invalid query with both sides of join being denoted as 'stream'. "
+          + "Stream-stream join is not yet supported. " + dumpRelPlanForNode(join));
+    }
+
+    if (isTablePosOnLeft && isTablePosOnRight) {
+      throw new SamzaException("Invalid query with both sides of join being denoted as 'table'. " +
+          dumpRelPlanForNode(join));
+    }
+
+    if (joinRelType.compareTo(JoinRelType.LEFT) == 0 && isTablePosOnLeft && !isTablePosOnRight) {
+      throw new SamzaException("Invalid query for outer left join. Left side of the join should be a 'stream' and "
+          + "right side of join should be a 'table'. " + dumpRelPlanForNode(join));
+    }
+
+    if (joinRelType.compareTo(JoinRelType.RIGHT) == 0 && isTablePosOnRight && !isTablePosOnLeft) {
+      throw new SamzaException("Invalid query for outer right join. Left side of the join should be a 'table' and "
+          + "right side of join should be a 'stream'. " + dumpRelPlanForNode(join));
+    }
+
+    validateJoinCondition(join.getCondition());
+  }
+
+  private void validateJoinCondition(RexNode operand) {
+    if (!(operand instanceof RexCall)) {
+      throw new SamzaException("SQL Query is not supported. Join condition operand " + operand +
+          " is of type " + operand.getClass());
+    }
+
+    RexCall condition = (RexCall) operand;
+
+    if (condition.isAlwaysTrue()) {
+      throw new SamzaException("Query results in a cross join, which is not supported. Please optimize the query."
+          + " It is expected that the joins should include JOIN ON operator in the sql query.");
+    }
+
+    if (condition.getKind() != SqlKind.EQUALS && condition.getKind() != SqlKind.AND) {
+      throw new SamzaException("Only equi-joins and AND operator is supported in join condition.");
+    }
+  }
+
+  // Get the stream and table indices corresponding to the fields given in the join condition by parsing through
+  // the condition.
+  private void getStreamAndTableKeyIds(List<RexNode> operands, final LogicalJoin join, boolean isTablePosOnRight,
+      List<Integer> streamIds, List<Integer> tableIds) {
```

Comment:
Are these streamKeyIds and tableKeyIds? Can you rename if that is the case, Streamid and tableId have a different meaning in Samza so it might be confusing.

Reply:
Sure